pkg/kv/kvserver/batcheval/cmd_export.go

```diff
@@ -225,18 +225,46 @@ func evalExport(
 		//    early exit and thus have a resume key despite
 		//    not having data.
 		if summary.DataSize == 0 {
-			if resume.Key != nil {
-				start = resume.Key
-				resumeKeyTS = resume.Timestamp
-				continue
-			} else {
-				break
+			hasResumeKey := resumeInfo.ResumeKey.Key != nil
+
+			// If we have a resumeKey, it means that we must have hit a resource
+			// constraint before exporting any data.
+			if hasResumeKey {
+				// If we hit our CPU limit we must return the response to the client
+				// instead of retrying immediately. This will give the scheduler a
+				// chance to move the goroutine off CPU allowing other processes to make
+				// progress. The client is responsible for handling pagination of
+				// ExportRequests.
+				if resumeInfo.CPUOverlimit {
+					// Note, since we have not exported any data we do not populate the
+					// `Files` field of the ExportResponse.
+					reply.ResumeSpan = &roachpb.Span{
+						Key:    resumeInfo.ResumeKey.Key,
+						EndKey: args.EndKey,
+					}
+					// TODO(during review): Do we want to add another resume reason
+					// specifically for CPU preemption.
+					reply.ResumeReason = roachpb.RESUME_UNKNOWN
+					break
+				} else {
+					// We should never come here. There should be no condition aside from
+					// resource constraints that results in an early exit without
+					// exporting any data. Regardless, if we have a resumeKey we
+					// immediately retry the ExportRequest from that key and timestamp
+					// onwards.
```

Comment:
Depending on our level of confident here, perhaps we could return an error in this case only in non-release builds and see if we ever trigger it in CI.